src/node-file-trace.ts

```diff
@@ -240,16 +266,18 @@ export class Job {
       return;
     }
 
-    if (Array.isArray(resolved)) {
-      for (const item of resolved) {
-        // ignore builtins
-        if (item.startsWith('node:')) return;
-        await this.emitDependency(item, path);
-      }
-    } else {
+    // For simplicity, force `resolved` to be an array
+    resolved = Array.isArray(resolved) ? resolved : [resolved];
+    for (let item of resolved) {
       // ignore builtins
-      if (resolved.startsWith('node:')) return;
-      await this.emitDependency(resolved, path);
+      if (item.startsWith('node:')) return;
+
+      // If querystring was stripped during resolution, restore it
+      if (queryString && !item.endsWith(queryString)) {
+        item += queryString;
+      }
```

Comment:
Can the query string exist on node built-ins?

Reply:
No, as it turns out. I did some testing:

(Note that this is all for ESM imports, since we know that CJS just treats `?` like any other character, and there are no built-ins with `?` in their name.)

- For node on its own:
  - Querystrings on relative imports (`import * as file1 from './file1.js?things'`) work (they get stripped).
  - Querystrings on regular node modules (`import * as semver from 'semver?stuff'`) error out with `Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'semver?stuff'`.
  - Querystrings on un-prefixed built-ins (`import * as fs from 'fs?morestuff'`) error out with `Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'fs?morestuff'`. (In other words, it doesn't recognize it as a built-in.)
  - Querystrings on prefixed built-ins (`import * as fs from 'node:fs?morestuff'`) error out with `Error [ERR_UNKNOWN_BUILTIN_MODULE]: No such built-in module: node:fs?morestuff`.

- For a webpack build:
  - Querystrings on relative imports (`import * as file1 from './file1.js?things'`) work (they get stripped).
  - Querystrings on regular node modules (`import * as semver from 'semver?stuff'`) work (they get stripped).
  - Querystrings on un-prefixed built-ins (`import * as fs from 'fs?morestuff'`) error out with `Module not found: Can't resolve 'fs?morestuff'`. (In other words, it doesn't recognize it as a built-in, strips it, and then errors when it doesn't find it in `node_modules`.)
  - Querystrings on prefixed built-ins (`import * as fs from 'node:fs?morestuff'`) error out with `Error [ERR_UNKNOWN_BUILTIN_MODULE]: No such built-in module: node:fs?morestuff`.

So not something we have to worry about. Might be worth saying so in a comment, though. I'll do that.